src/lib/journey/_utilities/metadata.utilities.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  isCbReadyByDefault,
+  isSelfSubmittingCb,
+  isStepSelfSubmittable,
+  requiresUserInput,
+} from '$journey/_utilities/step.utilities';
+
+import type { FRCallback } from '@forgerock/javascript-sdk';
+import type { CallbackMetadata, WidgetStep } from '$journey/journey.interfaces';
+
+export function buildStepMetadata(callbackMetadataArray: CallbackMetadata[]) {
+  return {
+    isStepSelfSubmittable: isStepSelfSubmittable(callbackMetadataArray),
+    numOfCallbacks: callbackMetadataArray.length,
+    numOfSelfSubmittableCbs: callbackMetadataArray.filter((cb) => !!cb.isSelfSubmittingCb).length,
+    numOfUserInputCbs: callbackMetadataArray.filter((cb) => !!cb.isUserInputRequired).length,
+  }
+}
+export function buildCallbackMetadata(
+  step: WidgetStep,
+  checkValidation: (callback: FRCallback) => boolean,
+) {
+  return step?.callbacks.map((callback, idx) => {
+    return {
+      isFirstInvalidInput: checkValidation(callback),
+      isReadyForSubmission: isCbReadyByDefault(callback),
+      isSelfSubmittingCb: isSelfSubmittingCb(callback),
+      isUserInputRequired: requiresUserInput(callback),
+      idx,
+    };
+  });
+}
```

Comment:
This is one of the main features of the "self-submitting" functionality that I added to support the `PollingWaitCallback` as well as the `ConfirmationCallback`. The idea is to build up these metadata objects outside of the actual step data and its callbacks, and use them to help inform portions of the code of the current step's nature.

I think this will also set us up for the new use of the Stage attribute by Platform UI. We can extract the data out of that stage properly, and inject it here to inform the components of intent.

There are currently two structures here: 1) the callback metadata, which is an array that matches the official callback array, and 2) the step metadata, which is a summary of the step itself and what it contains.

Reply:
This approach makes sense. I pulled in your code to run it, and found that we now see a reasonable amount of noise in the console when we run the app - and quite a lot more noise when we build it. These are warnings similar to the below, since we are declaring props that we are not using - these come from the new metadata and functions we are injecting.

```
16:11:34 [vite-plugin-svelte] /Users/david.adams/Checkouts/forgerock-web-login-framework-review/src/lib/journey/callbacks/username/validated-create-username.svelte:9:11 Validated_create_username has unused export property 'stepMetadata'. If it is for external reference only, please consider using `export const stepMetadata`
 7: export let callbackMetadata;
 8: export let selfSubmitFunction = null;
 9: export let stepMetadata;
```
Is there a way to reduce / suppress this noise, or are we happy to live with these for now? Not a huge problem, just one that would be good to reduce if it's easy enough since it can be unnerving seeing so many warnings!

Reply:
Yeah, I intent to solve these warnings, but I just ran out of time as this was supposed to be a 3 point story ... it ended up being about an 8. I could add addressing this in the CI story, maybe? But, I agree, I hate a noisy console log too :)